### PR TITLE
[Feature] Support loading data from Ceph in SOT dataset class

### PR DESCRIPTION
**Pull request by JingweiZhang12**
## Modification
This PR mainly contains the following:
1. loading data information files in text fromat from Ceph in SOT dataset class
2. loading images and annotations from Ceph in SOT dataset class

## Use cases (Optional)
If you want to load data from Ceph in SOT, you should modify your configs like the following:
```python
file_client_args = dict(
    backend='petrel',
    path_mapping=dict({
        'data/got10k':
        'openmmlab:s3://openmmlab/datasets/tracking/GOT10k',
        'data/trackingnet':
        'openmmlab:s3://openmmlab/datasets/tracking/TrackingNet',
        'data/lasot':
        'openmmlab:s3://openmmlab/datasets/tracking/LaSOT_full',
        'data/coco':
        'openmmlab:s3://openmmlab/datasets/detection/coco',
        'data/ILSVRC':
        'openmmlab:s3://openmmlab/datasets/tracking/ILSVRC',
        'data/otb100':
        'openmmlab:s3://openmmlab/datasets/tracking/OTB100',
        'data/UAV123':
        'openmmlab:s3://openmmlab/datasets/tracking/UAV123',
        'data/vot2018':
        'openmmlab:s3://openmmlab/datasets/tracking/VOT2018'
    }))

train_pipeline = [
    ...
    dict(
        type='LoadMultiImagesFromFile',
        to_float32=True,
        file_client_args=file_client_args),
    dict(
        type='SeqLoadAnnotations',
        with_bbox=True,
        with_label=False,
        file_client_args=file_client_args),
    ...
]

test_pipeline = [
    dict(
        type='LoadImageFromFile',
        to_float32=True,
        file_client_args=file_client_args),
    dict(
        type='LoadAnnotations',
        with_bbox=True,
        with_label=False,
        file_client_args=file_client_args),
    ...
]

data = dict(
    ...,
    train=dict(
        type='RandomSampleConcatDataset',
        dataset_sampling_weights=[1,1],
        dataset_cfgs=[
            dict(
                type='GOT10kDataset',
                ann_file=data_root +
                'got10k/annotations/got10k_train_infos.txt',
                img_prefix=data_root + 'got10k',
                pipeline=train_pipeline,
                split='train_vot',
                test_mode=False,
                file_client_args=file_client_args),
            dict(
                type='LaSOTDataset',
                ann_file=data_root + 'lasot/annotations/lasot_train_infos.txt',
                img_prefix=data_root + 'lasot/LaSOTBenchmark',
                pipeline=train_pipeline,
                split='train',
                test_mode=False,
                file_client_args=file_client_args),
           ]
        ...),
     val=dict(
        type='GOT10kDataset',
        ann_file=data_root + 'got10k/annotations/got10k_test_infos.txt',
        img_prefix=data_root + 'got10k',
        pipeline=test_pipeline,
        split='test',
        test_mode=True,
        file_client_args=file_client_args),
    test=dict(
        type='GOT10kDataset',
        ann_file=data_root + 'got10k/annotations/got10k_test_infos.txt',
        img_prefix=data_root + 'got10k',
        pipeline=test_pipeline,
        split='test',
        test_mode=True,
        file_client_args=file_client_args))
```

### CODE REVIEW

**Comment by JingweiZhang12** [comment]
I'll verify loading text-format information files from Ceph in trainin and test later.

File: mmtrack/datasets/got10k_dataset.py [hunk 1]
```diff
@@ -44,18 +44,19 @@ def load_data_infos(self, split='train'):
...
-                line = line.strip().replace('/', os.sep).split(',')
-                data_info = dict(
-                    video_path=line[0],
-                    ann_path=line[1],
-                    start_frame_id=int(line[2]),
-                    end_frame_id=int(line[3]),
-                    framename_template='%08d.jpg')
-                data_infos.append(data_info)
+        data_infos_str = self.file_client.get_text(
+            self.ann_file).strip().split('\n')
```
Comment (GT9505) [inline mmtrack/datasets/got10k_dataset.py:48]: Can we use the `self.loadtxt()` API?

**Review by GT9505 — COMMENTED** [review]
Please make sure that we support both `diskbackend` and `petrelbackend` in SOT dataset

File: mmtrack/datasets/base_sot_dataset.py [hunk 2]
```diff
@@ -43,6 +45,7 @@ def __init__(self,
                  test_mode=False,
                  bbox_min_size=0,
                  only_eval_visible=False,
+                 file_client_args=dict(backend='disk'),
```
Comment (GT9505) [inline mmtrack/datasets/base_sot_dataset.py:48]: miss the docstring for `file_client_args`

File: mmtrack/datasets/got10k_dataset.py (hunk 1 above)
Comment (JingweiZhang12) [reply mmtrack/datasets/got10k_dataset.py:48]: I don't use `self.loadtxt()` API for the efficiency.
`self.loadtxt()` is fit for the load of the whole txt files in the NumPy format.

Comment (JingweiZhang12) [reply mmtrack/datasets/got10k_dataset.py:48]: I have readjusted it.

File: mmtrack/datasets/sot_coco_dataset.py [hunk 3]
```diff
@@ -22,7 +23,13 @@ def __init__(self, ann_file, *args, **kwargs):
...
         """
-        self.coco = COCO(ann_file)
+        if 'file_client_args' in kwargs and kwargs['file_client_args'][
+                'backend'] != 'disk':
+            self.file_client = mmcv.FileClient(**kwargs['file_client_args'])
+            with self.file_client.get_local_path(ann_file) as local_path:
+                self.coco = COCO(local_path)
+        else:
+            self.coco = COCO(ann_file)
         super().__init__(*args, **kwargs)
```
Comment (GT9505) [inline mmtrack/datasets/sot_coco_dataset.py:33]: Too ugly here, we may only keep line 28-30?

File: mmtrack/datasets/sot_imagenet_vid_dataset.py [hunk 4]
```diff
@@ -21,7 +22,13 @@ def __init__(self, ann_file, *args, **kwargs):
...
                 `self.load_data_infos` function.
         """
-        self.coco = CocoVID(ann_file)
+        if 'file_client_args' in kwargs and kwargs['file_client_args'][
+                'backend'] != 'disk':
+            self.file_client = mmcv.FileClient(**kwargs['file_client_args'])
+            with self.file_client.get_local_path(ann_file) as local_path:
+                self.coco = CocoVID(local_path)
+        else:
+            self.coco = CocoVID(ann_file)
```
Comment (GT9505) [inline mmtrack/datasets/sot_imagenet_vid_dataset.py:31]: same here

**Comment by JingweiZhang12** [comment]
I have verified loading from Ceph for all SOT datasets. It's OK.

File: mmtrack/datasets/sot_coco_dataset.py (hunk 3 above)
Comment (JingweiZhang12) [reply mmtrack/datasets/sot_coco_dataset.py:33]: If only keep line 28-30, it will create a temporary file and read it even reading local files.

Comment (GT9505) [reply mmtrack/datasets/sot_coco_dataset.py:33]: I think it is acceptable. And mmdet also use [the same way](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/datasets/custom.py#L93-L95)

Comment (JingweiZhang12) [reply mmtrack/datasets/sot_coco_dataset.py:33]: OK， I modify it right now.

**Review by GT9505 — APPROVED** [review]

**Pull request merged**